File: meili_flutter_car_android/android/src/main/kotlin/com/meili/travel/flutter/car/MeiliParamsParser.kt
package com.flutter.meili

import com.meili.travel.api.AdditionalParams
import com.meili.travel.api.AvailParams
import com.meili.travel.api.DevelopmentEnvironment
import com.meili.travel.api.MeiliEnvironment
import com.meili.travel.api.MeiliFlow
import com.meili.travel.api.PreProductionEnvironment
import com.meili.travel.api.ProductionEnvironment
import com.meili.travel.api.UatEnvironment

internal fun parseEnv(envName: String?): MeiliEnvironment = when (envName?.lowercase()) {
    "prod", "production" -> ProductionEnvironment()
    "pre_prod", "preprod" -> PreProductionEnvironment()
    "uat" -> UatEnvironment()
    else -> DevelopmentEnvironment()
}

internal fun parseFlow(flowName: String?): MeiliFlow = when (flowName?.lowercase()) {
    "bookingmanager" -> MeiliFlow.BookingManager
    else -> MeiliFlow.Direct
}

internal fun parseAvailParams(map: Map<*, *>?): AvailParams? {
    map ?: return null
    return AvailParams(
        pickupLocation = map["pickupLocation"] as? String,
        dropoffLocation = map["dropoffLocation"] as? String,
        pickupDate = map["pickupDate"] as? String,
        pickupTime = map["pickupTime"] as? String,
        dropoffDate = map["dropoffDate"] as? String,
        dropoffTime = map["dropoffTime"] as? String,
        driverAge = (map["driverAge"] as? Number)?.toInt(),
        currencyCode = map["currencyCode"] as? String,
        residency = map["residency"] as? String,
        discountRequested = map["discountRequested"] as? Boolean,
        partnerLoyaltyAccountTier = map["partnerLoyaltyAccountTier"] as? String,
    )
}

@Suppress("UNCHECKED_CAST")
internal fun parseAdditionalParams(map: Map<*, *>?): AdditionalParams? {
    map ?: return null
    val supplierAccounts = map["supplierLoyaltyAccounts"] as? List<*>
    return AdditionalParams(
        passengerNameRecord = map["passengerNameRecord"] as? String,
        partnerLoyaltyAccount = map["partnerLoyaltyAccount"] as? String,
        partnerLoyaltyAccountTier = map["partnerLoyaltyAccountTier"] as? String,
        numberOfPassengers = (map["numberOfPassengers"] as? Number)?.toInt(),
        customerPartnerCode = map["customerPartnerCode"] as? String,
        flightNumber = map["flightNumber"] as? String,
        superPassengerNameRecord = map["superPassengerNameRecord"] as? String,
        infant = (map["infant"] as? Number)?.toInt(),
        child = (map["child"] as? Number)?.toInt(),
        teenager = (map["teenager"] as? Number)?.toInt(),
        supplierLoyaltyAccount = supplierAccounts?.firstOrNull()?.toString(),
        fareTypeAndFlex = map["fareTypeAndFlex"] as? String,
        departureAirport = map["departureAirport"] as? String,
        arrivalAirport = map["arrivalAirport"] as? String,
        ancillaryActivity = map["ancillaryActivity"] as? String,
        airlineFareAmount = (map["airlineFareAmount"] as? Number)?.toDouble(),
        airlineFareCurrency = map["airlineFareCurrency"] as? String,
        partnerCustomerID = map["partnerCustomerID"] as? String,
        firstName = map["firstName"] as? String,
        lastName = map["lastName"] as? String,
        email = map["email"] as? String,
        phoneNumbers = (map["phoneNumbers"] as? List<*>)?.map { it.toString() },
        companyName = map["companyName"] as? String,
        addressLine1 = map["addressLine1"] as? String,
        postCode = map["postCode"] as? String,
        city = map["city"] as? String,
        state = map["state"] as? String,
        confirmationId = map["confirmationId"] as? String,
        prefillOnly = map["prefillOnly"] as? Boolean,
        givenName = map["givenName"] as? String,
    )
}

File: meili_flutter_car_ios/ios/meili_flutter_car_ios/Sources/meili_flutter_car_ios/MeiliParamsParser.swift
//
//  MeiliParamsParser.swift
//  meili_flutter_ios
//

import Foundation
import MeiliSDK

func parseBookingParams(from dict: [String: Any]) -> AdditionalParams? {
    return AdditionalParams(
        partnerLoyaltyAccount: dict["partnerLoyaltyAccount"] as? String,
        partnerLoyaltyAccountTier: dict["partnerLoyaltyAccountTier"] as? String,
        numberOfPassengers: dict["numberOfPassengers"] as? Int,
        customerPartnerCode: dict["customerPartnerCode"] as? String,
        flightNumber: dict["flightNumber"] as? String,
        passengerNameRecord: dict["passengerNameRecord"] as? String,
        superPassengerNameRecord: dict["superPassengerNameRecord"] as? String,
        infant: dict["infant"] as? Int,
        child: dict["child"] as? Int,
        teenager: dict["teenager"] as? Int,
        supplierLoyaltyAccounts: dict["supplierLoyaltyAccount"] as? [String],
        fareTypeAndFlex: dict["fareTypeAndFlex"] as? String,
        departureAirport: dict["departureAirport"] as? String,
        arrivalAirport: dict["arrivalAirport"] as? String,
        ancillaryActivity: dict["ancillaryActivity"] as? String,
        airlineFareAmount: dict["airlineFareAmount"] as? Double,
        airlineFareCurrency: dict["airlineFareCurrency"] as? String,
        partnerCustomerID: dict["partnerCustomerID"] as? String,
        firstName: dict["firstName"] as? String,
        lastName: dict["lastName"] as? String ?? "",
        email: dict["email"] as? String,
        phoneNumbers: dict["phoneNumbers"] as? [Int],
        companyName: dict["companyName"] as? String,
        addressLine1: dict["addressLine1"] as? String,
        postCode: dict["postCode"] as? String,
        city: dict["city"] as? String,
        state: dict["state"] as? String,
        confirmationId: dict["confirmationId"] as? String ?? "",
        prefillOnly: dict["prefillOnly"] as? Bool
    )
}

// Every AvailParams field is optional on the SDK side. The Dart model makes most of them
// required, so hosts that only want to set the currency send "" and 0 to mean "not set".
func parseAvailParams(from dict: [String: Any]) -> AvailParams? {
    AvailParams(
        pickupLocation: dict.nonBlankString("pickupLocation"),
        dropoffLocation: dict.nonBlankString("dropoffLocation"),
        pickupDate: dict.nonBlankString("pickupDate"),
        pickupTime: dict.nonBlankString("pickupTime"),
        dropoffDate: dict.nonBlankString("dropoffDate"),
        dropoffTime: dict.nonBlankString("dropoffTime"),
        driverAge: (dict["driverAge"] as? Int).flatMap { $0 > 0 ? $0 : nil },
        currencyCode: dict.nonBlankString("currencyCode"),
        residency: dict.nonBlankString("residency"),
        discountRequested: dict["discountRequested"] as? Bool,
        partnerLoyaltyAccountTier: dict.nonBlankString("partnerLoyaltyAccountTier"),
        pickupDateTime: dict.nonBlankString("pickupDateTime"),
        dropoffDateTime: dict.nonBlankString("dropoffDateTime")
    )
}

private extension Dictionary where Key == String, Value == Any {
    func nonBlankString(_ key: String) -> String? {
        (self[key] as? String).flatMap { $0.trimmingCharacters(in: .whitespaces).isEmpty ? nil : $0 }
    }
}
